### ISSUE TITLE
Set SecurityContext for pods created by scan jobs

### ISSUE THREAD

**Issue by danielpacak** [aquasecurity/starboard#163]
To implement least privileged principle we should explicitly set SecurityContexts<sup>[1]</sup> on pod templates for each scan job. Such SecurityContext should reflect minimal privileges to run a given scanner. The goal here is to review required permissions and set them explicitly for each pod template comprising a scan job descriptor.

[1]: https://kubernetes.io/docs/tasks/configure-pod-container/security-context/

**Comment by akath19** [comment]
@danielpacak I'd like to pick this one up, here's my suggested SecurityContext:

Pod Level:
```
securityContext:
   runAsUser: 1900
   runAsGroup: 1900
   fsGroup: 1900
   runAsNonRoot: true
```

Container level:

```
securityContext:
  allowPrivilegeEscalation: false
  capabilities:
      drop:
       - all
  privileged: false
  readOnlyRootFilesystem: false
````

I've tested this with trivy & polaris jobs, I haven't tested this with kube-bench/hunter jobs yet.

**Comment by danielpacak** [comment]
👋 @akath19 Happy to assign this one to you!

I think we can start with hardcoded structure literals to specify SecurityContexts for scan jobs. Later on we can think of making it configurable.

For the SecurityContext itself please go ahead and open a PR so I can review it. I believe we cannot run Trivy as non root user, but if that's the case our integration should catch it.

**Comment by danielpacak** [comment]
@akath19 Any update or blocker on this one? Shall I keep it assigned to you?

**Comment by akath19** [comment]
Sorry about that, I've been super busy at work.

I already have a basic PR for this, I'll send it soon

**Comment by danielpacak** [comment]
> Sorry about that, I've been super busy at work.
> 
> I already have a basic PR for this, I'll send it soon

No worries and no rush. It's just a ping. Feel free to split this issue into smaller PRs, e.g. for each scanner. That way we can make it less overwhelming and simpler to review.

**Comment by lizrice** [comment]
Hi @akath19, we're unassigning this for now so that someone else could pick it up. If your PR is ready or nearly ready and you can submit it soon please let us know and we'll assign it back to you. Thanks!

**Comment by mlevesquedion** [comment]
I would be interested in picking this up.

**Comment by danielpacak** [comment]
That's great @mlevesquedion . Please go ahead and consider splitting this task into smaller pull requests, e.g. review SecurityContext for each scanner one by one.

**Comment by mlevesquedion** [comment]
I have been trying to run `kube-hunter` as a non-root user, but I have not succeeded so far. Even when I give all capabilities and privileged is set to `true`, the hunt fails, apparently because of this error:

```python
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/kube_hunter/core/events/handler.py", line 137, in worker
    hook.execute()
  File "/usr/local/lib/python3.8/site-packages/kube_hunter/modules/discovery/hosts.py", line 112, in execute
    subnets = self.traceroute_discovery()
  File "/usr/local/lib/python3.8/site-packages/kube_hunter/modules/discovery/hosts.py", line 146, in traceroute_discovery
    node_internal_ip = srp1(
  File "/usr/local/lib/python3.8/site-packages/scapy/sendrecv.py", line 564, in srp1
    ans, _ = srp(*args, **kargs)
  File "/usr/local/lib/python3.8/site-packages/scapy/sendrecv.py", line 552, in srp
    s = conf.L2socket(promisc=promisc, iface=iface,
  File "/usr/local/lib/python3.8/site-packages/scapy/arch/linux.py", line 398, in __init__
    self.ins = socket.socket(socket.AF_PACKET, socket.SOCK_RAW, socket.htons(type))  # noqa: E501
  File "/usr/local/lib/python3.8/socket.py", line 231, in __init__
    _socket.socket.__init__(self, family, type, proto, fileno)
PermissionError: [Errno 1] Operation not permitted
```

I will need to dig deeper into this.

I have had better luck with `kube-bench`, PTAL:

**Comment by mlevesquedion** [comment]
I took a deeper look at `kube-hunter`.

Fundamentally, the reason that adding `CAP_NET_RAW` to the `securityContext` is not enough is that doing this merely sets the container's `inheritable` and `bounding` capability sets. Since the kernel checks `effective` capabilities, and `kubernetes` does not support `ambient` capabilities, in order to actually be able to use the capabilities, either the binary needs to have the capabilities set on it using e.g. `sudo setcap cap_net_raw+ep <binary>` (since `kube-hunter` is a Python script, in this case the "binary" is actually the `python` interpreter) or the script needs to be capabilities-aware.

However, it appears that since `v0.4.1`, `kube-hunter` [no longer needs CAP_NET_RAW](https://github.com/aquasecurity/kube-hunter/pull/416). Using `v0.4.1`, I was able to run `kube-hunter` as a non-root user with no capabilities. The only difference in output that I could observe is that since `CAP_NET_RAW` was unavailable, `kube-hunter` did not report that it was enabled.

Therefore, I think we should bump the default image ref to use `v0.4.1`, and only set the security context if the image version is at least `0.4.1`.

WDYT @danielpacak?

**Comment by danielpacak** [comment]
> I took a deeper look at `kube-hunter`.
> 
> Fundamentally, the reason that adding `CAP_NET_RAW` to the `securityContext` is not enough is that doing this merely sets the container's `inheritable` and `bounding` capability sets. Since the kernel checks `effective` capabilities, and `kubernetes` does not support `ambient` capabilities, in order to actually be able to use the capabilities, either the binary needs to have the capabilities set on it using e.g. `sudo setcap cap_net_raw+ep <binary>` (since `kube-hunter` is a Python script, in this case the "binary" is actually the `python` interpreter) or the script needs to be capabilities-aware.
> 
> However, it appears that since `v0.4.1`, `kube-hunter` [no longer needs CAP_NET_RAW](https://github.com/aquasecurity/kube-hunter/pull/416). Using `v0.4.1`, I was able to run `kube-hunter` as a non-root user with no capabilities. The only difference in output that I could observe is that since `CAP_NET_RAW` was unavailable, `kube-hunter` did not report that it was enabled.
> 
> Therefore, I think we should bump the default image ref to use `v0.4.1`, and only set the security context if the image version is at least `0.4.1`.
> 
> WDYT @danielpacak?

I asked @danielsagi  for advice how to proceed with this one. Let's see what he says.

**Comment by danielsagi** [comment]
Red

> I took a deeper look at `kube-hunter`.
> 
> Fundamentally, the reason that adding `CAP_NET_RAW` to the `securityContext` is not enough is that doing this merely sets the container's `inheritable` and `bounding` capability sets. Since the kernel checks `effective` capabilities, and `kubernetes` does not support `ambient` capabilities, in order to actually be able to use the capabilities, either the binary needs to have the capabilities set on it using e.g. `sudo setcap cap_net_raw+ep <binary>` (since `kube-hunter` is a Python script, in this case the "binary" is actually the `python` interpreter) or the script needs to be capabilities-aware.
> 
> However, it appears that since `v0.4.1`, `kube-hunter` [no longer needs CAP_NET_RAW](https://github.com/aquasecurity/kube-hunter/pull/416). Using `v0.4.1`, I was able to run `kube-hunter` as a non-root user with no capabilities. The only difference in output that I could observe is that since `CAP_NET_RAW` was unavailable, `kube-hunter` did not report that it was enabled.
> 
> Therefore, I think we should bump the default image ref to use `v0.4.1`, and only set the security context if the image version is at least `0.4.1`.
> 
> WDYT @danielpacak?

Hi, very detailed research was going on here! 
Love this.
Indeed until version v0.4.1in kube-hunter we relied on CAP_NET_RAW for scanning. That's why I released v0.4.1. 

Cap_net_raw is not needed anymore from this version. So as I see it if you bump the version to v0.4.1 you can freely remove the Cap. This will not affect kube-hunter in any way.

**Comment by chen-keinan** [comment]
@mikalai-ivanisenka Starboard is no longer maintained , All effort has ben shift to [trivy-operator](https://github.com/aquasecurity/trivy-operator)  with lot more capabilities and enhancements
